### PR TITLE
Forbid most Transactional methods on OnDemand extensions. 

### LINKED ISSUE #416: Using Transactional with onDemand makes it easy to do nonsensical things
From the mailing list:

> For transaction support I've currently chosen to implement the Transactional interface and manually call dao.begin(), dao.commit() and dao.rollback() and my UserResource has a few transactional and a few non-transactional methods:

``` java
public interface UserDao implements Transactional<UserDao> { 
}

public class UserResource {

   public void insertUser(User user) {
      userDao.insertUser(user);
   }

   public void updateUser(User user) {
      try {
         userDao.begin();
         userDao.updateUser(user);
         userDao.commit(); 
      catch (Exception e) {
         userDao.rollback();
      }
   }
}
```

This usage doesn't work, but it's not clear to an outsider, and we should make it blow up or harder to misuse.

### PR DESCRIPTION
**Pull request by stevenschlansker**
Fixes #416

### CODE REVIEW

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/TransactionalHelper.java [hunk 1]
```diff
@@ -26,13 +27,13 @@
...
+            h.put(Transactional.class.getMethod("begin"), guard(new BeginHandler()));
+            h.put(Transactional.class.getMethod("commit"), guard(new CommitHandler()));
+            h.put(Transactional.class.getMethod("rollback"), guard(new RollbackHandler()));
 
-            h.put(Transactional.class.getMethod("savepoint", String.class), new SavepointHandler());
-            h.put(Transactional.class.getMethod("releaseSavepoint", String.class), new ReleaseSavepointHandler());
-            h.put(Transactional.class.getMethod("rollbackToSavepoint", String.class), new RollbackSavepointHandler());
+            h.put(Transactional.class.getMethod("savepoint", String.class), guard(new SavepointHandler()));
+            h.put(Transactional.class.getMethod("releaseSavepoint", String.class), guard(new ReleaseSavepointHandler()));
+            h.put(Transactional.class.getMethod("rollbackToSavepoint", String.class), guard(new RollbackSavepointHandler()));
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/TransactionalHelper.java]: The `rollback`, `savepoint`, `releaseSavepoint`, and `rollbackToSavepoint` methods should all be valid within an `inTransaction` or `useTransaction` call.

File: core/src/main/java/org/jdbi/v3/core/OnDemandExtensions.java [hunk 2]
```diff
@@ -72,6 +75,11 @@
         return extensionType.cast(
                 Proxy.newProxyInstance(
                         extensionType.getClassLoader(),
-                        new Class[]{extensionType}, handler));
+                        new Class<?>[] { extensionType }, handler));
+    }
+
+    public static boolean inOnDemandExtension()
+    {
+        return IN_ON_DEMAND_EXTENSION.get();
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/OnDemandExtensions.java]: I think we need something more closely scoped to the instance.

Example: if the on-demand instance has a default method that calls a `@CreateSqlObject` that creates a `Transactional` type, then calls transactional methods internally, that should be valid usage.

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/mixins/TestTransactionalOnDemand.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        txn.rollbackToSavepoint("somewhere");
+    }
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void testReleaseSavepoint() throws Exception
+    {
+        txn.releaseSavepoint("somewhere");
+    }
+
+    public interface Txn extends Transactional<Txn>
```
Comment (qualidafial) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/mixins/TestTransactionalOnDemand.java]: I'd like to see tests showing that `inTransaction` and `useTransaction` still work as expected.

**Comment by qualidafial** [comment]
A simpler solution might be to add a check in `LazyHandleSupplier.close()`: before closing the handle, check if the handle is in a transaction. If so, rollback the transaction, close the handle, and throw an exception with an informative message.

This should catch users who call `Handle.begin()` without an accompanying `commit()` or `rollback()`, as well as those who call `Transactional.begin()` on an on-demand SQL object.

**Comment by qualidafial** [comment]
@stevenschlansker You mind if I take a stab at this? I have an approach in mind that should be pretty unobtrusive.

**Comment by stevenschlansker** [comment]
Go for it.

**Comment by qualidafial** [comment]
Opened PR #597

**Pull request closed without merging**